### DISCUSSION TITLE
a problem of multivariable

### DISCUSSION THREAD

**Discussion by dingqiushi** [idaholab/moose#21922] [Q&A General] [answered]
I want to write a kernel for multivariable, but I have some problems with the data type when calculating the residuals and Jacobi.
1. In a kernel suitable for single variable, I know that the component of the data type `RealVectorValue` is of type Real, so in a `kernel `suitable for multiple variables, what component of the data type is of type `RealEigenVector`? In other words, what is the data type of a variable consisting of the `RealEigenVector` type?
2. Because of the member function type requirement, the data type of the return value must be `RealEigenVector`. By testing I know that the data type of _grad_test[_i][_qp] is `RealGradient`, and I want to know what data type can be multiplied with the `RealGradient` type and the result of the multiplication is` RealEigenVector` type. I tried the `RealVectorArrayValue` type and the `RealEigenMatrix` type ,the following errors were reported at compile time.
（1)
```
error: no match for operator* (operand types are ’libMesh::RealVectorArrayValue {aka Eigen::Matrix<double,-1,3>}” and 'const value_type {aka const libMesh::VectorValue<double>}') 
return flux * _test[_i][_qp] + (lift*_grad_test[_i][_qp]);
```
（2）
```
error: no match for operator* (operand types are ’libMesh::RealEigenMatrix {aka Eigen::Matrix<double,-1,-1>}” and 'const value_type {aka const libMesh::VectorValue<double>}') 
return flux * _test[_i][_qp] + (lift*_grad_test[_i][_qp]);
```
3. In my attempts to pass values, I found that it actually compiles when the left side of the equal sign is of type `RealGradient` and the right side is of type Real, does this make sense?

**Comment by GiudGiud** [comment]
Hello

So you want a kernel that contributes to the residual for multiple variables? Is this an array variable or multiple regular variables?

Variables with the RealEigenVector type are array variables. 

Guillaume

**Reply by dingqiushi** [reply]
Yes, the variables I use are array variables. The kernel inherit from ArrayKernel.C.

**Comment by YaqiWang** [comment, marked as answer]
All definitions are in `moose/framework/include/utils/MooseTypes.h` including the following:
```
typedef Eigen::Matrix<Real, Eigen::Dynamic, 1> RealEigenVector;
```
You also want to take a look at `ArrayDiffusion.C`.

**Reply by dingqiushi** [reply]
Thanks for the replies, I noticed ArrayDiffusion.C but I didn't see anything similar to what I did and didn't find any variables to multiply with _grad_test[_i][_qp].
I have one more question for you all here, what is the difference between _array_grad_test and _grad_test?

**Reply by GiudGiud** [reply]
Regular variables multiply with grad_test when doing integration by parts on a laplacian for example.

array_grad_test is the equivalent for array variables

**Reply by dingqiushi** [reply]
Is there a data type which can fit multiple variables and at the same time represent a four-dimensional matrix?
Tip: The four-dimensional matrix I want to construct (size 5*5*3*3) can be understood as a large matrix (5*5) with many small matrices (25 small 3*3 matrices in total) nested inside it.

**Reply by GiudGiud** [reply]
no there are no tensor variables in moose. You may use a very large array if it's convenient with regards to the equations

**Reply by dingqiushi** [reply]
Thanks for your reply, I saw this definition in the Eigen library and it seems to represent a four dimensional variable, is there a similar expression in moose?
![859b95ebfaa2a4edf7ac2ad77d4ec5c](https://user-images.githubusercontent.com/108042711/186557060-329b7fbd-b179-42c9-b6be-71ea114af2b5.png)

http://eigen.tuxfamily.org/dox/classEigen_1_1Array.html#title4

**Reply by GiudGiud** [reply]
You can use these constructs but not for variables currently. You could create tensor material properties for example
